### PR TITLE
microsoft_gsl: 2.1.0 -> 3.1.0

### PR DESCRIPTION
**Pull request by NickCao**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 130527` at fe195f61 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>microsoft_gsl</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/libraries/microsoft_gsl/default.nix:35:3:
```
   |
35 |   checkPhase = "make test";
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/libraries/microsoft_gsl/default.nix [hunk 1]
```diff
@@ -1,44 +1,49 @@
...
+  nativeBuildInputs = [ cmake pkg-config ];
+  buildInputs = [ gtest ];
 
-  installPhase = ''
-    mkdir -p $out/include
-    mv ../include/ $out/
-  '';
+  # check phase just runs the unit tests, so skip it if
+  # we're doing a cross build
+  doCheck = stdenv.hostPlatform == stdenv.buildPlatform;
```
Comment (Mic92) [inline pkgs/development/libraries/microsoft_gsl/default.nix:34]: Is this not the default for checks anyway?

File: pkgs/development/libraries/microsoft_gsl/default.nix [hunk 2]
```diff
@@ -1,44 +1,49 @@
...
-    sha256 = "09f08lxqm00152bx9yrizlgabzpzxlpbv06h00z4w78yxywgxlgx";
+    sha256 = "sha256-gIpyuNlp3mvR8r1Azs2r76ElEodykRLGAwMN4BDJez0=";
   };
 
-  # build phase just runs the unit tests, so skip it if
-  # we're doing a cross build
-  nativeBuildInputs = [ catch cmake ];
-  buildPhase = if nativeBuild then "make" else "true";
+  patches = [
+    # Search for GoogleTest via pkg-config first, ref: https://github.com/NixOS/nixpkgs/pull/130525
```
Comment (Mic92) [inline pkgs/development/libraries/microsoft_gsl/default.nix:22]: Not sure how https://github.com/NixOS/nixpkgs/pull/130525 relate to this.

File: pkgs/development/libraries/microsoft_gsl/default.nix (hunk 1 above)
Comment (NickCao) [reply pkgs/development/libraries/microsoft_gsl/default.nix:34]: Was not aware of that 😂

**Review by risicle — APPROVED** [review]
LGTM, `nixpkgs-review` happy macos 10.15, nixos x86_64.

**Pull request merged**